tests/test_context.py

```diff
@@ -101,6 +104,120 @@ def test_get_trace_empty(self):
+            self.assertIsNone(trace)
+            self.assertIsNone(sampled)
+
+    def test_partial_flush_too_many(self):
+        """
+        When calling `Context.get`
+            When partial flushing is enabled
+            When we have more than the minimum number of spans needed to flush
+                We return the finished spans
+        """
+        tracer = get_dummy_tracer()
+        ctx = Context()
+
+        # Create a root span with 5 children, all of the children are finished, the root is not
+        root = Span(tracer=tracer, name='root')
+        ctx.add_span(root)
+        for i in range(5):
+            child = Span(tracer=tracer, name='child_{}'.format(i), trace_id=root.trace_id, parent_id=root.span_id)
+            child._parent = root
+            child.finish()
+            ctx.add_span(child)
+            ctx.close_span(child)
+
+        with override_config('context', dict(partial_flush_enabled=True, partial_flush_min_spans=1)):
+            trace, sampled = ctx.get()
+
+        self.assertIsNotNone(trace)
+        self.assertIsNotNone(sampled)
+
+        self.assertEqual(len(trace), 5)
+        self.assertEqual(
+            set(['child_0', 'child_1', 'child_2', 'child_3', 'child_4']),
+            set([span.name for span in trace])
+        )
+
+        # Ensure we clear/reset internal stats as expected
+        self.assertEqual(ctx._finished_spans, 0)
+        self.assertEqual(ctx._trace, [root])
+        with override_config('context', dict(partial_flush_enabled=True, partial_flush_min_spans=5)):
+            trace, sampled = ctx.get()
+            self.assertIsNone(trace)
+            self.assertIsNone(sampled)
+
+    def test_partial_flush_too_few(self):
+        """
+        When calling `Context.get`
+            When partial flushing is enabled
+            When we do not have enough finished spans to flush
+                We return no spans
+        """
+        tracer = get_dummy_tracer()
+        ctx = Context()
+
+        # Create a root span with 5 children, all of the children are finished, the root is not
+        root = Span(tracer=tracer, name='root')
+        ctx.add_span(root)
+        for i in range(5):
+            child = Span(tracer=tracer, name='child_{}'.format(i), trace_id=root.trace_id, parent_id=root.span_id)
+            child._parent = root
+            child.finish()
+            ctx.add_span(child)
+            ctx.close_span(child)
+
+        # Test with having 1 too few spans for partial flush
+        with override_config('context', dict(partial_flush_enabled=True, partial_flush_min_spans=6)):
+            trace, sampled = ctx.get()
+
+        self.assertIsNone(trace)
+        self.assertIsNone(sampled)
+
+        self.assertEqual(len(ctx._trace), 6)
+        self.assertEqual(ctx._finished_spans, 5)
+        self.assertEqual(
+            set(['root', 'child_0', 'child_1', 'child_2', 'child_3', 'child_4']),
+            set([span.name for span in ctx._trace])
+        )
+
```

Comment:
We should make sure to test when there are unfinished spans that they remain in the trace.